src/bottom_up_iterator.jl

```diff
@@ -0,0 +1,223 @@
+"""
+	abstract type BottomUpIterator <: ProgramIterator
+
+Enumerates programs in a bottom-up fashion. This means that it starts with the smallest programs and gradually builds up to larger programs.
+The exploration of the search space is done by making use of the priority function, which associates each program with its cost.
+
+Concrete implementations of this iterator should implement the following methods:
+- `order(iter::BottomUpIterator, grammar::ContextSensitiveGrammar)::Vector{Int64}`: Returns the order in which the rules should be enumerated.
+- `pick(iter::BottomUpIterator, grammar::ContextSensitiveGrammar, state::BottomUpState, rule::Int64)::Vector{RuleNode}`: Returns the programs that can be created by applying the given rule.
+- `priority_function(iter::BottomUpIterator, grammar::ContextSensitiveGrammar, program::RuleNode, state::BottomUpState)::Int64`: Returns the priority of the given program.
+"""
+abstract type BottomUpIterator <: ProgramIterator end
+
+Base.@doc """
+    @programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+
+A basic implementation of the bottom-up iterator. It will enumerate all programs in increasing order based on their depth.
+""" DepthIterator
+@programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
```

Comment:
`DepthIterator` is a specific instance of bottom-up search where the programs are ordered according to their depth?